### PR TITLE
Add React Native perf metrics behind flag

### PR DESCRIPTION
**Pull request by EdmondChuiHW**
# Summary

This PR enables performance metrics to be collected in **internal builds** of CDT Frontend for performance, reliability, and efficiency.

* Adds new flag for React Native perf metrics
  * Default: off, i.e. no-op for open-source
  * Will only be enabled for internal builds, e.g. via a custom `embedderScript.js`
* To support setting these configurations before the entry points, `embedderScript.js` is now a `defer` script.
  * As `type="module"` script is `defer` by definition, the order of these script tags in HTML will be the order of execution per [spec](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#defer).

With the flag enabled, you can use `addListener` to handle events:

```js
const unsubscribeFn = RNPerfMetrics.Impl.getInstance().addListener(event => console.log(event));
```

# Planned in next PRs

* Add event type definitions
* Add callsites for `sendEvent(…)` (no-op when flag is false/missing)

# Test plan

In both release and default builds:

```js
// in embedderScript.js
globalThis.enableReactNativePerfMetrics = true;
globalThis.enableReactNativePerfMetricsGlobalPostMessage = true;

window.addEventListener('message', event => {
  console.log('got em', event.data.tag);
});
```

```js
// after rn_inspector.ts
RNPerfMetrics.Impl.getInstance().sendEvent({'tag': 'TestyMcTest'});
```

Expect console to print `TestyMcTest`

# Upstreaming plan

- [ ] This commit should be sent as a patch to the upstream `devtools-frontend` repo. I've reviewed the [contribution guide](https://docs.google.com/document/d/1WNF-KqRSzPLUUfZqQG5AFeU_Ll8TfWYcJasa_XGf7ro/edit#heading=h.9kj7femz1xg5).
- [x] This commit is React Native-specific and cannot be upstreamed.

# References

* Similar PR in React DevTools: https://github.com/facebook/react/pull/22276
* Upstream `UserMetrics`: https://github.com/facebookexperimental/rn-chrome-devtools-frontend/blob/9370b35e1979ec6953ff59ed78c8751e4f463750/front_end/core/host/UserMetrics.ts#L34

### CODE REVIEW

File: front_end/core/host/RNPerfMetrics.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
+// Copyright 2024 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (hoxyq) [inline front_end/core/host/RNPerfMetrics.ts:8]: I think you need to add this new file (and other) to https://github.com/facebookexperimental/rn-chrome-devtools-frontend/blob/8c5e07a93900cdfc87174c3b82908ff1d75cf79f/scripts/eslint_rules/lib/check_license_header.js#L69-L72 in order to get correct headers

**Review by hoxyq — APPROVED** [review]
Looks good, please see the comment about code paths for Meta's license headers

File: front_end/core/host/RNPerfMetrics.ts (hunk 1 above)
Comment (EdmondChuiHW) [reply front_end/core/host/RNPerfMetrics.ts]: Ah. Some eslint rule was looking for `addEventListener` and expects a second param. That crashed so the license linter didn't run either

File: scripts/eslint_rules/lib/check_license_header.js [hunk 2]
```diff
@@ -69,6 +69,7 @@ const EXCLUDED_FILES = [
 const META_CODE_PATHS = [
   'entrypoints/rn_inspector',
   'panels/rn_welcome',
+  'core/host/RNPerfMetrics.ts',
```
Comment (hoxyq) [inline scripts/eslint_rules/lib/check_license_header.js:72]: @EdmondChuiHW consider adding `front_end/global_typings/react_native.d.ts` here as well in one of the next PRs.

Comment (EdmondChuiHW) [reply scripts/eslint_rules/lib/check_license_header.js:72]: Hmm. Looks like `.d.ts` are excluded somehow. The upstream file here doesn't have license and the linter doesn't complain.

https://github.com/facebookexperimental/rn-chrome-devtools-frontend/blob/9370b35e1979ec6953ff59ed78c8751e4f463750/front_end/global_typings/request_idle_callback.d.ts#L1

Added ours to the list for clarity in https://github.com/facebookexperimental/rn-chrome-devtools-frontend/pull/18

**Pull request merged**